edge-util/src/Microsoft.Azure.Devices.Edge.Storage/DbStoreProviderWithBackupRestore.cs

```diff
@@ -0,0 +1,351 @@
+            await provider.RestoreAsync();
+            return provider;
+        }
+
+        private async Task RestoreAsync()
+        {
+            string backupMetadataFilePath = Path.Combine(this.backupPath, BackupMetadataFileName);
+
+            if (!File.Exists(backupMetadataFilePath))
+            {
+                this.events.NoBackupsForRestore();
+                return;
+            }
+
+            try
+            {
+                string fileText = File.ReadAllText(backupMetadataFilePath);
+                BackupMetadataList backupMetadataList = JsonConvert.DeserializeObject<BackupMetadataList>(fileText);
+                BackupMetadata backupMetadata = backupMetadataList.Backups[0];
+                this.events.BackupInformation(backupMetadata.Id, backupMetadata.SerializationFormat, backupMetadata.TimestampUtc, backupMetadata.Stores);
+
+                string latestBackupDirPath = Path.Combine(this.backupPath, backupMetadata.Id.ToString());
+                if (Directory.Exists(latestBackupDirPath))
+                {
+                    this.events.RestoringFromBackup(backupMetadata.Id);
+                    foreach (string store in backupMetadata.Stores)
+                    {
+                        IDbStore dbStore;
+                        if (!store.Equals(DefaultStoreBackupName, StringComparison.OrdinalIgnoreCase))
+                        {
+                            dbStore = this.GetDbStore(store);
+                        }
+                        else
+                        {
+                            dbStore = this.GetDbStore();
+                        }
+
+                        await this.dbStoreBackupRestore.RestoreAsync(store, dbStore, latestBackupDirPath);
+                    }
+
+                    this.events.RestoreComplete();
+                }
+                else
+                {
+                    this.events.NoBackupsForRestore();
+                }
+            }
+            catch (Exception exception)
+            {
+                this.events.RestoreFailure($"The restore operation failed with error ${exception}.");
+
+                // Clean up any restored state in the dictionary if the backup fails midway.
+                foreach (string store in this.dbStores.ToList())
+                {
+                    this.RemoveDbStore(store);
+                }
+
+                this.RemoveDbStore();
+            }
+            finally
+            {
+                // Delete all other backups as we've either:
+                // 1. Restored successfully.
+                // 2. Failed during restore which would indicate a bad backup.
+                this.CleanupAllBackups(this.backupPath);
+            }
+        }
+
+        public override IDbStore GetDbStore(string partitionName)
+        {
+            this.dbStores.Add(partitionName);
```

Comment:
This should be a concurrent collection

Reply:
Changed to ConcurrentDictionary with a single dummy byte as value